### PR TITLE
Feature-310: Migrating the Bash Steps to Python

### PR DESCRIPTION
**Pull request by andsimakov**
### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
- Migrated bash scripts to Python for test execution.

- Added Docker integration for building and running tests.

- Updated dependencies to include Docker and dotenv support.

- Enhanced documentation to reflect Python-based test execution.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>run_test_all.py</strong><dd><code>Python script for running all tests with Docker</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests_integration/run_test_all.py

<li>Added Python script to replace bash for running all tests.<br> <li> Integrated Docker for building and running test containers.<br> <li> Defined multiple test configurations for various programming <br>languages.<br> <li> Implemented logging and error handling for Docker operations.

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/qodo-cover/pull/315/files#diff-6d10384f955e5b2887a83711df2ee5a8c9b895b7fe96f717f1d08caaf5e12462">+229/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>run_test_with_docker.py</strong><dd><code>Python script for running individual tests with Docker</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests_integration/run_test_with_docker.py

<li>Added Python script to replace bash for running individual tests.<br> <li> Implemented Docker image building and container management.<br> <li> Included functionality for copying files and executing commands in <br>containers.<br> <li> Enhanced error handling for Docker operations.

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/qodo-cover/pull/315/files#diff-41cfd74a1effaa3b584fdd513b8fdd3969d20bbeb3c43678fbd769f4d4ad3bfa">+202/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Dockerfile</strong><dd><code>Update Dockerfile for Python-based test setup</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Dockerfile

<li>Updated Dockerfile to use <code>make setup-installer</code> for setup.<br> <li> Adjusted commands to align with new Python-based test scripts.

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/qodo-cover/pull/315/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>.env</strong><dd><code>Add environment variables for test configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests_integration/.env

- Added environment variables for model, iterations, and coverage.

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/qodo-cover/pull/315/files#diff-0cb990657b2ec7dcd8894c7716cc507f8953e757255bd443df6321641d44386e">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Update dependencies for Docker and dotenv support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pyproject.toml

<li>Added <code>python-dotenv</code> and <code>docker</code> dependencies.<br> <li> Reorganized dependencies for better clarity.

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/qodo-cover/pull/315/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Update README for Python-based test execution</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests_integration/README.md

<li>Updated documentation to reflect Python-based test execution.<br> <li> Replaced bash script examples with Python script examples.<br> <li> Added instructions for using different LLMs.

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/qodo-cover/pull/315/files#diff-300591ffc7d29907a4d44ca8ca1f5ab7ab38b6494cdd775c84cec717df7f7b4c">+14/-14</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by EmbeddedDevops1** [comment]
@andsimakov Thanks for working on this issue. I was able to get it up and running using WSL on Windows (looks like it now supports the Docker pip package). I had to make some modifications to the `--run-installer` option to get it working. 

Attached is the git patch: [fix-docker-runner.patch](https://github.com/user-attachments/files/19495837/fix-docker-runner.patch). Please download it into the repo and run `git apply fix-docker-runner.patch` then push your changes once more.

**Comment by EmbeddedDevops1** [comment]
@andsimakov after you've applied that patch there are some other issues that need to be ironed out. Here's what a local run looks like on my machine:

> $ poetry run python tests_integration/run_test_all.py
> 2025-03-27 16:46:37,429 - run_test_with_docker - INFO - Pulling the Docker image embeddeddevops/c_cli:latest...
> 2025-03-27 16:46:39,026 - run_test_with_docker - INFO - Removing existing container with name cover-agent-container...
> 2025-03-27 16:46:40,189 - run_test_with_docker - INFO - Running command in container: /usr/local/bin/cover-agent --source-file-path calc.c --test-file-path test_calc.c --code-coverage-report-path coverage_filtered.info --test-command sh build_and_test_with_coverage.sh --coverage-type lcov --desired-coverage 50 --max-iterations 4 --strict-coverage --model gpt-4o-2024-11-20
> Streaming results from LLM model...
> ```yaml
> language: c
> testing_framework: Unity
> number_of_tests: 1
> test_headers_indentation: 0
> ```
> 
> Streaming results from LLM model...
> ```yaml
> language: c
> testing_framework: Unity
> number_of_tests: 1
> relevant_line_number_to_insert_tests_after: 9
> relevant_line_number_to_insert_imports_after: 2
> ```
> 
> Streaming results from LLM model...
> ```yaml
> language: c
> testing_framework: Unity
> number_of_tests: 1
> relevant_line_number_to_insert_tests_after: 9
> relevant_line_number_to_insert_imports_after: 2
> ```2025-03-27 23:46:46,401 - cover_agent.UnitTestValidator - INFO - Running build/test command to generate coverage report: "sh build_and_test_with_coverage.sh"
> 2025-03-27 23:46:47,077 - cover_agent.UnitTestValidator - INFO - Initial coverage: 12.5%
> 2025-03-27 23:46:47,077 - cover_agent.CoverAgent - INFO - Current Coverage: 12.5%
> 2025-03-27 23:46:47,077 - cover_agent.CoverAgent - INFO - Desired Coverage: 50%
> 
> 
> Streaming results from LLM model...
> ```yaml
> language: c
> existing_test_function_signature: |
>   void test_Addition(void);
> new_tests:
> - test_behavior: |
>     Test that the function correctly subtracts two numbers when the operation is "--subtract".
>   test_name: |
>     test_Subtraction
>   test_code: |
>     void test_Subtraction(void) {
>         int status;
>         double result = calculate(10, 4, "--subtract", &status);
>         TEST_ASSERT_EQUAL_INT(6, result);
>         TEST_ASSERT_EQUAL_INT(0, status);
>     }
>   new_imports_code: |
>     ""
>   test_tags: happy path
> 
> - test_behavior: |
>     Test that the function correctly multiplies two numbers when the operation is "--multiply".
>   test_name: |
>     test_Multiplication
>   test_code: |
>     void test_Multiplication(void) {
>         int status;
>         double result = calculate(7, 3, "--multiply", &status);
>         TEST_ASSERT_EQUAL_INT(21, result);
>         TEST_ASSERT_EQUAL_INT(0, status);
>     }
>   new_imports_code: |
>     ""
>   test_tags: happy path
> 
> - test_behavior: |
>     Test that the function handles division by zero when the operation is "--divide".
>   test_name: |
>     test_DivisionByZero
>   test_code: |
>     void test_DivisionByZero(void) {
>         int status;
>         double result = calculate(10, 0, "--divide", &status);
>         TEST_ASSERT_EQUAL_INT(0, result);
>         TEST_ASSERT_EQUAL_INT(1, status);
>     }
>   new_imports_code: |
>     ""
>   test_tags: edge case
> 
> - test_behavior: |
>     Test that the function returns an error when an unknown operation is provided.
>   test_name: |
>     test_UnknownOperation
>   test_code: |
>     void test_UnknownOperation(void) {
>         int status;
>         double result = calculate(5, 3, "--unknown", &status);
>         TEST_ASSERT_EQUAL_INT(0, result);
>         TEST_ASSERT_EQUAL_INT(2, status);
>     }
>   new_imports_code: |
>     ""
>   test_tags: edge case
> ```2025-03-27 23:46:52,619 - cover_agent.UnitTestValidator - INFO - Running test with the following command: "sh build_and_test_with_coverage.sh"
> 2025-03-27 23:46:53,153 - cover_agent.UnitTestValidator - INFO - Test passed and coverage increased. Current coverage: 18.75%
> 2025-03-27 23:46:53,153 - cover_agent.UnitTestValidator - INFO - Running test with the following command: "sh build_and_test_with_coverage.sh"
> 2025-03-27 23:46:53,707 - cover_agent.UnitTestValidator - INFO - Test passed and coverage increased. Current coverage: 25.0%
> 2025-03-27 23:46:53,707 - cover_agent.UnitTestValidator - INFO - Running test with the following command: "sh build_and_test_with_coverage.sh"
> 2025-03-27 23:46:54,292 - cover_agent.UnitTestValidator - INFO - Test passed and coverage increased. Current coverage: 37.5%
> 2025-03-27 23:46:54,292 - cover_agent.UnitTestValidator - INFO - Running test with the following command: "sh build_and_test_with_coverage.sh"
> 2025-03-27 23:46:54,860 - cover_agent.UnitTestValidator - INFO - Test passed and coverage increased. Current coverage: 50.0%
> 2025-03-27 23:46:54,930 - cover_agent.UnitTestValidator - INFO - Running build/test command to generate coverage report: "sh build_and_test_with_coverage.sh"
> 2025-03-27 23:46:55,522 - cover_agent.UnitTestValidator - INFO - Initial coverage: 50.0%
> 2025-03-27 23:46:55,522 - cover_agent.CoverAgent - INFO - Reached above target coverage of 50% (Current Coverage: 50.0%) in 0 iterations.
> 2025-03-27 23:46:55,522 - cover_agent.CoverAgent - INFO - Total number of input tokens used for LLM model gpt-4o-2024-11-20: 3073
> 2025-03-27 23:46:55,522 - cover_agent.CoverAgent - INFO - Total number of output tokens used for LLM model gpt-4o-2024-11-20: 569
> 
> 
> 2025-03-27 16:47:06,445 - run_test_with_docker - INFO - Pulling the Docker image embeddeddevops/cpp_cli:latest...
> 2025-03-27 16:47:08,317 - run_test_with_docker - INFO - Running command in container: /usr/local/bin/cover-agent --source-file-path calculator.cpp --test-file-path test_calculator.cpp --code-coverage-report-path coverage.xml --test-command sh build_and_test_with_coverage.sh --coverage-type cobertura --desired-coverage 70 --max-iterations 3 --strict-coverage --model gpt-4o-2024-11-20
> Streaming results from LLM model...
> ```yaml
> language: c++
> testing_framework: googletest
> number_of_tests: 1
> test_headers_indentation: 0
> ```
> 
> Streaming results from LLM model...
> ```yaml
> language: c++
> testing_framework: Google Test
> number_of_tests: 1
> relevant_line_number_to_insert_tests_after: 6
> relevant_line_number_to_insert_imports_after: 2
> ```2025-03-27 23:47:14,223 - cover_agent.UnitTestValidator - INFO - Running build/test command to generate coverage report: "sh build_and_test_with_coverage.sh"
> Traceback (most recent call last):
>   File "cover_agent/main.py", line 149, in <module>
>   File "cover_agent/main.py", line 145, in main
>   File "cover_agent/CoverAgent.py", line 315, in run
>   File "cover_agent/CoverAgent.py", line 203, in init
>   File "cover_agent/UnitTestValidator.py", line 133, in get_coverage
>   File "cover_agent/UnitTestValidator.py", line 303, in run_coverage
> AssertionError: Fatal: Error running test command. Are you sure the command is correct? "sh build_and_test_with_coverage.sh"
> Exit code 1. 
> Stdout: 
> Compiling the test application...
>  
> Stderr: 
> cmake: /tmp/_MEI8Z0i8J/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by cmake)
> cmake: /tmp/_MEI8Z0i8J/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by cmake)
> cmake: /tmp/_MEI8Z0i8J/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /lib/x86_64-linux-gnu/libjsoncpp.so.25)
> cmake: /tmp/_MEI8Z0i8J/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /lib/x86_64-linux-gnu/libicuuc.so.72)
> 
> [PYI-13:ERROR] Failed to execute script 'main' due to unhandled exception!
> 
> 
> 2025-03-27 16:47:25,167 - run_test_with_docker - INFO - Pulling the Docker image embeddeddevops/csharp_webservice:latest...
> 2025-03-27 16:47:27,369 - run_test_with_docker - INFO - Running command in container: /usr/local/bin/cover-agent --source-file-path CalculatorApi/CalculatorController.cs --test-file-path CalculatorApi.Tests/CalculatorControllerTests.cs --code-coverage-report-path CalculatorApi.Tests/TestResults/coverage.cobertura.xml --test-command dotnet test --collect:"XPlat Code Coverage" CalculatorApi.Tests/ && find . -name "coverage.cobertura.xml" -exec mv {} CalculatorApi.Tests/TestResults/coverage.cobertura.xml \; --coverage-type cobertura --desired-coverage 50 --max-iterations 3 --strict-coverage --model gpt-4o-2024-11-20
> Streaming results from LLM model...
> ```yaml
> language: c#
> testing_framework: xunit
> number_of_tests: 1
> test_headers_indentation: 4
> ```
> 
> Streaming results from LLM model...
> ```yaml
> language: c#
> testing_framework: xUnit
> number_of_tests: 1
> relevant_line_number_to_insert_tests_after: 20
> relevant_line_number_to_insert_imports_after: 3
> ```2025-03-27 23:47:33,116 - cover_agent.UnitTestValidator - INFO - Running build/test command to generate coverage report: "dotnet test --collect:"XPlat Code Coverage" CalculatorApi.Tests/ && find . -name "coverage.cobertura.xml" -exec mv {} CalculatorApi.Tests/TestResults/coverage.cobertura.xml \;"
> Traceback (most recent call last):
>   File "cover_agent/main.py", line 149, in <module>
>   File "cover_agent/main.py", line 145, in main
>   File "cover_agent/CoverAgent.py", line 315, in run
>   File "cover_agent/CoverAgent.py", line 203, in init
>   File "cover_agent/UnitTestValidator.py", line 133, in get_coverage
>   File "cover_agent/UnitTestValidator.py", line 303, in run_coverage
> AssertionError: Fatal: Error running test command. Are you sure the command is correct? "dotnet test --collect:"XPlat Code Coverage" CalculatorApi.Tests/ && find . -name "coverage.cobertura.xml" -exec mv {} CalculatorApi.Tests/TestResults/coverage.cobertura.xml \;"
> Exit code 134. 
> Stdout: 
>  
> Stderr: 
> Process terminated. Couldn't find a valid ICU package installed on the system. Please install libicu (or icu-libs) using your package manager and try again. Alternatively you can set the configuration flag System.Globalization.Invariant to true if you want to run with no globalization support. Please see https://aka.ms/dotnet-missing-libicu for more information.
>    at System.Environment.FailFast(System.String)
>    at System.Globalization.GlobalizationMode+Settings..cctor()
>    at System.Globalization.CultureData.CreateCultureWithInvariantData()
>    at System.Globalization.CultureData.get_Invariant()
>    at System.Globalization.CultureInfo..cctor()
>    at System.Globalization.CultureInfo.get_CurrentUICulture()
>    at System.TimeZoneInfo.GetUtcStandardDisplayName()
>    at System.TimeZoneInfo.CreateUtcTimeZone()
>    at System.TimeZoneInfo..cctor()
>    at System.DateTime.get_Now()
>    at Microsoft.DotNet.Cli.Program.Main(System.String[])
> Aborted (core dumped)
> 
> [PYI-13:ERROR] Failed to execute script 'main' due to unhandled exception!
> 
> 
> 2025-03-27 16:47:44,631 - run_test_with_docker - INFO - Pulling the Docker image embeddeddevops/go_webservice:latest...
> 2025-03-27 16:47:46,904 - run_test_with_docker - INFO - Running command in container: /usr/local/bin/cover-agent --source-file-path app.go --test-file-path app_test.go --code-coverage-report-path  --test-command go test -coverprofile=coverage.out && gocov convert coverage.out | gocov-xml > coverage.xml --coverage-type cobertura --desired-coverage 70 --max-iterations 4 --strict-coverage --model gpt-4o-2024-11-20
> Streaming results from LLM model...
> ```yaml
> language: go
> testing_framework: testify
> number_of_tests: 1
> test_headers_indentation: 0
> ```
> 
> Streaming results from LLM model...
> ```yaml
> language: go
> testing_framework: testify
> number_of_tests: 1
> relevant_line_number_to_insert_tests_after: 20
> relevant_line_number_to_insert_imports_after: 9
> ```2025-03-27 23:47:52,258 - cover_agent.UnitTestValidator - INFO - Running build/test command to generate coverage report: "go test -coverprofile=coverage.out && gocov convert coverage.out | gocov-xml > coverage.xml"
> 2025-03-27 23:47:55,940 - cover_agent.UnitTestValidator - ERROR - Error in coverage processing: Fatal: Coverage report "" was not generated.
> Traceback (most recent call last):
>   File "cover_agent/main.py", line 149, in <module>
>   File "cover_agent/main.py", line 145, in main
>   File "cover_agent/CoverAgent.py", line 315, in run
>   File "cover_agent/CoverAgent.py", line 203, in init
>   File "cover_agent/UnitTestValidator.py", line 133, in get_coverage
>   File "cover_agent/UnitTestValidator.py", line 308, in run_coverage
>   File "cover_agent/UnitTestValidator.py", line 764, in post_process_coverage_report
>   File "cover_agent/CoverageProcessor.py", line 56, in process_coverage_report
>   File "cover_agent/CoverageProcessor.py", line 69, in verify_report_update
> AssertionError: Fatal: Coverage report "" was not generated.
> [PYI-13:ERROR] Failed to execute script 'main' due to unhandled exception!

## Issues
So far I see two major issues:
1. The commands we are inputting don't seem to be processed correctly. This makes sense because you're going from command line --> Python --> Docker (and whatever other layers in between). I think you may need to handle these as "raw" strings (e.g. `r"These will all be printed: \;>"`) since we're using all sorts of funky characters like `>`, `;`, etc.
2. When an error occurs with a single container run we should exit the whole script and also make sure the containers get removed. This is an important "cleanup process" that happens in `tests_integration/test_all.sh`/`tests_integration/test_with_docker.sh` to ensure we don't waste compute time (and LLM token fees) on integration testing.

**Comment by andsimakov** [comment]
> @andsimakov Thanks for working on this issue. I was able to get it up and running using WSL on Windows (looks like it now supports the Docker pip package). I had to make some modifications to the `--run-installer` option to get it working (partially my fault for not updating the Dockerfile).
> 
> While I was at it, I added some extra streaming capabilities for debugging purposes. Perhaps in a follow up MR we can just create a class called something like `docker_utils.py` that abstracts this all away from `run_test_all.py` and `run_test_with_docker.py`.
> 
> Attached is the git patch: [fix-docker-runner.patch](https://github.com/user-attachments/files/19495912/fix-docker-runner.patch). Please download it into the repo and run `git apply fix-docker-runner.patch` then push your changes once more.

File: tests_integration/run_test_all.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Args for running tests with Docker.")
+    parser.add_argument("--model", default=constants.MODEL, help="Which LLM model to use.")
+    parser.add_argument("--run-installer", action="store_true", help="Build `cover-agent` binary.")
+    args = parser.parse_args()
+
+    model = args.model
+    run_installer = args.run_installer
```
Comment (EmbeddedDevops1) [inline tests_integration/run_test_all.py:31]: @andsimakov Can we get rid of the installer piece and and just use a script such as build_installer.sh?
```bash
#!/bin/bash

## Utility script to build the installer for the cover-agent project
# This should be run from the root of the repository

set -e  # Exit immediately if a command exits with a non-zero status
set -o pipefail  # Exit if any command in a pipeline fails
# set -x  # Print commands and their arguments as they are executed

echo "Building cover-agent-installer..."
# Build the installer within a Docker container
docker build --platform linux/amd64 -t cover-agent-installer -f Dockerfile .

# Create dist directory if it doesn't exist
mkdir -p dist

# Run the container and mount the dist directory
docker run --platform linux/amd64 --rm --volume "$(pwd)/dist:/app/dist" cover-agent-installer
```

Alternatively we can do this in a follow up PR if you'd like

Comment (andsimakov) [reply tests_integration/run_test_all.py]: It's done. Review it, please.

File: templated_tests/csharp_webservice/Dockerfile [hunk 2]
```diff
@@ -6,4 +6,4 @@ WORKDIR /app
 COPY . /app
 
 # Restore dependencies
-RUN dotnet restore CalculatorApi.Tests/CalculatorApi.Tests.csproj
```
Comment (EmbeddedDevops1) [inline templated_tests/csharp_webservice/Dockerfile:9]: No need for this change anymore. Run `git checkout origin/main -- templated_tests/csharp_webservice/Dockerfile`, commit, and push to revert this change.

File: tests_integration/README.md [hunk 3]
```diff
@@ -11,30 +11,30 @@ You will also need [Docker](https://www.docker.com/) installed.
...
 
 ## How to Run
 You can run these example test suites using a locally hosted LLM or in the cloud just as you would normally with Cover Agent.
 
 ### Running the Tests
 To run the full test suite, simply run the following command from the root of the repository:
-```
-sh tests_integration/test_all.sh
+```shell
+python tests_integration/run_test_all.py
```
Comment (EmbeddedDevops1) [inline tests_integration/README.md:22]: Instruction should be `poetry run python ...` versus `python` since you included all the necessary packages in the `pyproject.toml`.

File: tests_integration/docker_utils.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
+import io
```
Comment (EmbeddedDevops1) [inline tests_integration/docker_utils.py:1]: @andsimakov Can you add docstrings and inline comments to all of these functions?

File: tests_integration/constants.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+from enum import Enum
```
Comment (EmbeddedDevops1) [inline tests_integration/constants.py:1]: I'm not sure this file is needed. Can we just place it in `run_test_all.py`? Some of these settings will be placed into a configuration class when #299 is completed.

File: tests_integration/constants.py
Comment (EmbeddedDevops1) [inline tests_integration/constants.py:1]: I'm not sure this file is needed. Can we just place it in `run_test_all.py`? Some of these settings will be placed into a configuration class when #299 is completed.

File: tests_integration/docker_utils.py
Comment (EmbeddedDevops1) [inline tests_integration/docker_utils.py:1]: @andsimakov Can you add docstrings and inline comments to all of these functions?

File: tests_integration/run_test_with_docker.py
Comment (EmbeddedDevops1) [inline tests_integration/run_test_with_docker.py:1]: Can you add inline comments and docstrings to this file?

File: tests_integration/README.md (hunk 3 above)
Comment (andsimakov) [reply tests_integration/README.md]: All the occurrences are fixed.

File: tests_integration/docker_utils.py
Comment (andsimakov) [reply tests_integration/docker_utils.py:1]: It's done.

File: templated_tests/csharp_webservice/Dockerfile (hunk 2 above)
Comment (EmbeddedDevops1) [reply templated_tests/csharp_webservice/Dockerfile:9]: Did you rebase beforehand? I’m still seeing changes here…

File: tests_integration/run_test_with_docker.py
Comment (andsimakov) [reply tests_integration/run_test_with_docker.py:1]: It's done.
